### ISSUE TITLE
New session cannot be destroyed

### ISSUE THREAD

**Issue by oldium** [bungle/lua-resty-session#175]
Session created with `local s = require "resty.session".start()` cannot be directly destroyed with `s:destroy()`. It fails with

```
/usr/local/openresty/luajit/share/lua/5.1/resty/session.lua:2077: unable to destroy nonexistent or closed session
```

Such session is in state `new`, not `open`, which is checked in the `destroy()` call.

This is problem to libraries like lua-resty-openidc, which can get a started session from the application and cannot determine if calling to `s:destroy()` is safe or not.

**Comment by bungle** [comment]
@oldium I think there are two ways to fix this: you call `:save` if session was inexistent OR we call `:save` when the session was new here: https://github.com/bungle/lua-resty-session/blob/master/lib/resty/session.lua#L2726.

Perhaps `:destroy` could be a noop when there is nothing to destroy.

**Comment by bungle** [comment]
I think why we skipped the `save` on `start` when the session was inexistent was that you rarely want to have session that has no data. But making `:destroy` a `noop` may be an option for `new` sessions (that have not been saved yet).

**Comment by oldium** [comment]
There is a real issue in [lua-resty-openidc](https://github.com/zmartzone/lua-resty-openidc/pull/489#issuecomment-1843287528) library. The library accepts either a session configuration (library opens the session) or an opened session (application opens the session). Moreover, in case of logout URI the library is responsible for destroying the session.

So if the application decides to always pass-in the opened session, it might happen that when user without cookies visits directly the logout URI, the application opens the new session when calling lua-resty-openidc library to handle the call, and the library tries to destroy the session, because this is what it should do on the session at logout URI.

There is no session API to know that the session is non-existent (freshly created and not-yet saved) and cannot be destroyed. This is only known to the application by checking the return value of `session.start`, but this is unknown to the library.

It would be good to behave as no-op in this particular case (destroy on new not-yet-saved session). I think this is a reasonable requirement - simply speaking, you should be able to destroy an opened session.

You could also enhance the session API to add the information that the session was new and not yet saved. Anyway, it makes no sense to me to check if it is valid to call destroy on the session, this should be part of the destroy logic.

This worked in 3.x, but stopped working in 4.x, so for me this is a regression.

**Comment by bungle** [comment]
>  it might happen that when user without cookies visits directly the logout URI

Why aren't you then just calling:
```
require("resty.session").destroy()
```

**Comment by bungle** [comment]
> is not even part of the session instance API

Yes, that is what I am thinking. There is `session.state` which reports the state but it is not documented. This is other way to solve this.

In theory you could just do:
```
if session.state == "OPEN" then
  session:destroy()
